src/projection.jl

```diff
@@ -0,0 +1,77 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    projector([T::Type], x)
+
+Returns a `project(dx)` closure which maps `dx` onto type `T`, such that it is the
+same size as `x`. If `T` is not provided, it is assumed to be the type of `x`.
+
+It's necessary to have `x` to ensure that it's possible to project e.g. `AbstractZero`s
+onto `Array`s -- this wouldn't be possible with type information alone because the neither
+`AbstractZero`s nor `T` know what size of `Array` to produce.
+"""
+function projector end
+
+projector(x) = projector(typeof(x), x)
+
+# fallback (structs)
+function projector(::Type{T}, x::T) where T
+    project(dx::T) = dx
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Numbers
+function projector(::Type{T}, x::T) where {T<:Real}
+    project(dx::Real) = T(dx)
+    project(dx::Number) = T(real(dx)) # to avoid InexactError
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
```

Comment:
I wonder if there should be some `struct Project` which is returned, in part to avoid writing these out every time.

Reply:
Could you clarify how this would work?

Reply:
One attempt is here: https://gist.github.com/mcabbott/8a84086cc604d34b5e8dff2eb3839f3a